python/tank/context.py

```diff
@@ -1208,7 +1279,7 @@ def _build_clean_entity(ent):
     if fallback_to_ctx_from_entity:
         # entity dict doesn't contain enough information to build a 
         # safe, valid context so fall back on 'from_entity':
-        return from_entity(tk, entity_type, entity_id)
+        return from_entity(tk, original_entity_type, original_entity_id)
```

Comment:
I think the code might have been simpler if you simply assigned a source_entity in the original `if published file` above and the called a new method called `_from_entity(tk, entity_type, entity_id, original_entity`.

Then you can update `from_entity` to call `_from_entity(tk, entity_type, entity_id, None)`. Would that have simplified the changes and still work?